picoquic/packet.c

```diff
@@ -34,110 +34,30 @@
 #include <stdlib.h>
 #include <string.h>
 
-/**
- * If the version is not know, need special code as long as we support
- * draft-20 and the corresponding format */
-
-int picoquic_is_old_header_invariant(
-    picoquic_quic_t* quic,
-    uint8_t* bytes,
-    size_t length,
-    picoquic_packet_header* ph)
+int picoquic_is_header_version_supported(uint32_t version)
 {
-    int ret = 0; /* return 1 if old invariant */
-    if ((ph->vn & 0xFFFFFF00) == 0xFF000000) {
+    int ret = 0; /* return 1 if version is supported */
+
+    if ((version & 0xFFFF0000) == 0) {
+        /* Final versions (0x0001-0xFFFF) and version negotiation (0x0) */
+        ret = 1;
+    } else if ((version & 0xFFFFFF00) == 0xFF000000 && (version & 0xFF) > 20) {
         /* Draft versions before #20 use the old invariants */
-        int draft_nb = ph->vn & 0xFF;
-        ret = (draft_nb <= 20) ? 1 : 0;
+        ret = 1;
+    } else if ((version & 0x0F0F0F0F) == 0x0A0A0A0A) {
+        /* version greasing, i.e. force version negotiation */
+        ret = 1;
+    } else if (
+        version == PICOQUIC_INTERNAL_TEST_VERSION_1 ||
+        version == PICOQUIC_INTERNAL_TEST_VERSION_2) {
+        /* Internal versions use the new invariants */
+        ret = 1;
     }
-    else if (ph->vn != 0 && (
```

Comment:
I would personally just remove this entirely. The protocol defines the version independent invariants that constrain all future versions. The change in May 2019 was pretty much the last chance of fixing something. After that, once some deployment has started, that becomes impossible. Facebook, Cloudflare and Google are already shipping products with the new format, so change is pretty much impossible.

Reply:
👍